### PR TITLE
Refactor `tuftool` to use `RepositoryEditor`

### LINKED ISSUE #43: Move functionality out of tuftool into a library
When it comes to interacting and updating TUF repos, quite a bit of useful functionality is still contained in `tuftool` and isn't usable by outside code. This functionality should be put in a library.

Right off the bat I can see a few functions that should be split out:
* `tuftool/src/create.rs : write_metadata()` : This function takes a role, computes its SHA/length, signs it, and writes it to the TUF output directory
* `tuftool/src/create.rs : process_target()` : Given a `&Path`, this function computes its SHA/length, builds a `Target` object, and writes the file to the appropriate TUF output directory.
*  `tuftool/src/source.rs : KeySource.read()` : Fetching / parsing signing keys is necessary to sign a repo.
* `tuftool/src/key.rs: keys_for_root()` : Given a signing key, get the appropriate root keys.
* `tuftool/src/key.rs : sign_metadata_inner()` : This function does some crucial parsing `root.json` to get the proper keys to use to actually sign.

### PR DESCRIPTION
**Pull request by zmrow**
*Issue #, if available:*
Fixes #43 

*Description of changes:*
The tl;dr is: "simplify `tuftool` by using `RepositoryEditor`."

The actual list of changes is:
* Add "getters" to `SignedRole` so a user can access, but not change data.
* Use `RepositoryEditor` anywhere a repository is being manipulated inside `tuftool`. This includes the `create` and `refresh` commands.
* **Breaking change**: Move the `sign` command to a subcommand of `tuftool root`. This functionality is primarily meant of creation of the `root.json` anyway.
* **Breaking change**: `tuftool create` now *symlinks* targets to the given directory rather than *copying* them. This is a better use of disk space as repositories or targets get larger. *If we still want to keep the copy functionality we can provide a flag the user can pass?*
* Remove any unused files
* Remove any unused error variants

*Testing done:*
* All unit and integration tests pass (which gives me pretty good confidence that we're creating repos the right way)
* Manually create a repository locally; targets are named correctly and properly symlinked.
* Manually refresh a repository locally; everything looks good here as well.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tuftool/src/error.rs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
+// Not really worried about the memory penalty of large enum variants here
+#![allow(clippy::large_enum_variant)]
```
Comment (tjkirch) [inline tuftool/src/error.rs:5]: What makes it large?  Nothing particularly stands out to me, and it seems smaller than it did before...

File: tuftool/src/root.rs [hunk 2]
```diff
@@ -85,6 +87,13 @@ pub(crate) enum Command {
         #[structopt(short = "r", long = "role")]
         roles: Vec<RoleType>,
     },
+    /// Set the signature count threshold for a role
+    Sign {
```
Comment (tjkirch) [inline tuftool/src/root.rs]: I think the comment was copypasta :)

File: tuftool/src/root.rs [hunk 3]
```diff
@@ -262,6 +272,21 @@ impl Command {
...
+
+        let signed_root = SignedRole::new(
+            root.signed.clone(),
+            &root.signed,
+            &[key_source],
+            &SystemRandom::new(),
+        )
+        .context(error::SignRoot { path })?;
+
+        std::fs::write(&path, signed_root.buffer()).context(error::FileWrite { path: &path })?;
```
Comment (tjkirch) [inline tuftool/src/root.rs]: Would be pretty funny to partially overwrite someone's root key with half of the signed data :D  Perhaps we should write to a temp file and move it over the original path?  (nit)

File: tuftool/src/create.rs [hunk 4]
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
...
+            .timestamp_expires(self.timestamp_expires);
+
+        let signed_repo = editor.sign(&self.keys).context(error::SignRepo)?;
+
+        let metadata_dir = &self.outdir.join("metadata");
+        let targets_dir = &self.outdir.join("targets");
+        signed_repo.write(metadata_dir).context(error::WriteRepo)?;
+        signed_repo
+            .link_targets(&self.indir, targets_dir)
+            .context(error::LinkTarget {
```
Comment (tjkirch) [inline tuftool/src/create.rs]:
```suggestion
            .context(error::LinkTargets {
```

File: tuftool/src/error.rs [hunk 5]
```diff
@@ -253,6 +254,12 @@ pub(crate) enum Error {
         backtrace: Backtrace,
     },
 
+    #[snafu(display("Failed writing repo data to disk: {}", source))]
+    WriteRepo {
```
Comment (tjkirch) [inline tuftool/src/error.rs:284]: Can we include the directory where we were trying to write the repo?

File: tuftool/src/create.rs [hunk 6]
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
 
 impl CreateArgs {
     pub(crate) fn run(&self) -> Result<()> {
-        if let Some(jobs) = self.jobs {
-            rayon::ThreadPoolBuilder::new()
```
Comment (tjkirch) [inline tuftool/src/create.rs:81]: rayon didn't seem to actually be used in refresh.rs, so I understand that removal, but I think it's used here in create to parallelize building targets, and I think we're losing that with the move to the repo editor... right?  Does it take a lot longer to create repos?  I wonder if we should include that as an improvement to repo editor in another commit for this PR?

File: tuftool/src/create.rs [hunk 7]
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
...
+            .snapshot_expires(self.snapshot_expires)
+            .timestamp_version(self.timestamp_version)
+            .timestamp_expires(self.timestamp_expires);
+
+        let signed_repo = editor.sign(&self.keys).context(error::SignRepo)?;
+
+        let metadata_dir = &self.outdir.join("metadata");
+        let targets_dir = &self.outdir.join("targets");
+        signed_repo.write(metadata_dir).context(error::WriteRepo)?;
+        signed_repo
```
Comment (tjkirch) [inline tuftool/src/create.rs]: Should we link targets before we write metadata so the metadata is valid/usable as soon as it's there?

File: tuftool/src/create.rs [hunk 8]
```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
...
-        let mut length = 0;
-        loop {
-            match file.read(&mut buf).context(error::FileRead { path })? {
-                0 => break,
-                n => {
-                    digest.update(&buf[..n]);
-                    length += n as u64;
-                }
+    // Build a list of files from the given directory
+    fn build_targets(&self) -> Result<Vec<PathBuf>> {
```
Comment (tjkirch) [inline tuftool/src/create.rs]:
```suggestion
    fn find_targets(&self) -> Result<Vec<PathBuf>> {
```
Not really building anymore...

**Review by tjkirch — COMMENTED** [review]
Look at all that beautiful red!

File: tuftool/src/error.rs (hunk 1 above)
Comment (zmrow) [reply tuftool/src/error.rs:5]: Clippy is angry that some of the enum variants are larger than others. I really think it's overly pedantic in this case and for a CLI tool I don't think we care about this.
```
warning: large size difference between variants
   --> tuftool/src/error.rs:145:5
    |
145 | /     LinkTarget {
146 | |         indir: PathBuf,
147 | |         outdir: PathBuf,
148 | |         source: tough::error::Error,
149 | |         backtrace: Backtrace,
150 | |     },
    | |_____^
```

File: tuftool/src/root.rs (hunk 2 above)
Comment (zmrow) [reply tuftool/src/root.rs]: Oops

File: tuftool/src/root.rs (hunk 3 above)
Comment (zmrow) [reply tuftool/src/root.rs]: Good point. :) I'll fix that.

File: tuftool/src/create.rs (hunk 7 above)
Comment (zmrow) [reply tuftool/src/create.rs]: I'll change it as you're right, but I can't see there being a race condition of any kind here unless a user is using their local machine as a "real" repo and polling for updates constantly. That being said... we have no idea what a user is going to do. :)

**Comment by zmrow** [comment]
Fixes @tjkirch 's comments except for the multithreading question and the clippy comment.

**Comment by zmrow** [comment]
@tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?

**Comment by tjkirch** [comment]
> @tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?

You know I like it, I requested it :)  A flag would be nice, I guess, but I'm not sure of a reason for it yet.

File: tuftool/src/create.rs (hunk 6 above)
Comment (tjkirch) [reply tuftool/src/create.rs:81]: Per offline discussion, tough shouldn't own the parallelism, but tuftool could/should.  If we can add `Target`s individually to the editor, rather than just paths, then we can build `Target`s in parallel in tuftool and add them to the editor as they're ready.  I don't think it has to be done here, but it does feel unfortunate to lose the parallelism with the change...

**Comment by zmrow** [comment]
> > @tjkirch @iliana : This PR symlinks the targets to the given outdir rather than copying them. How do you feel about that? Should we provide a flag so the user can choose to copy the targets?
> 
> You know I like it, I requested it :) A flag would be nice, I guess, but I'm not sure of a reason for it yet.

Unless @iliana has different feelings, I'll leave it out for now.

File: tuftool/src/create.rs (hunk 6 above)
Comment (webern) [reply tuftool/src/create.rs:81]: We should probably add Windows to the build matrix. Easy to do, seems relevant with respect to symlinks, unless they 'just work' the same over there.

**Review by webern — APPROVED** [review]
Wow, now I see why the editor is a good thing!

**Comment by zmrow** [comment]
The above commit adds the threading support back into `tuftool create`. I ended up being able to use a decent amount of the original code, so I got that going for me too, which is nice.

**Comment by zmrow** [comment]
Rebased onto `develop`

File: tough/src/editor/mod.rs [hunk 9]
```diff
@@ -189,8 +189,21 @@ impl RepositoryEditor {
         Ok(self)
     }
 
+    /// Add a `Target` to the repository
+    pub fn add_target(&mut self, name: String, target: Target) -> Result<&mut Self> {
```
Comment (tjkirch) [inline tough/src/editor/mod.rs]: It won't let me comment below, but we should be able to use this at the end of `add_target_path` now.

File: tough/src/editor/test.rs [hunk 10]
```diff
@@ -58,19 +58,23 @@ mod tests {
...
     #[test]
-    fn add_targets() {
+    fn add_targets_from_multiple_sources() {
         let targets: Signed<Targets> = serde_json::from_str(include_str!(
             "../../tests/data/tuf-reference-impl/metadata/targets.json"
         ))
         .unwrap();
-        let target3 = targets_path().join("file3.txt");
+        let target3_path = targets_path().join("file3.txt");
+        let target2_path = targets_path().join("file2.txt");
```
Comment (tjkirch) [inline tough/src/editor/test.rs:67]: Hmm, should we name this something other than what's in the existing targets?  Or maybe test both, replacing an existing target and adding a new one from a `Target`?

File: tuftool/src/refresh.rs [hunk 11]
```diff
@@ -75,124 +68,43 @@ pub(crate) struct RefreshArgs {
 
 impl RefreshArgs {
     pub(crate) fn run(&self) -> Result<()> {
-        if let Some(jobs) = self.jobs {
```
Comment (tjkirch) [inline tuftool/src/refresh.rs:78]: It won't let me comment on the new location of this, but it made me think - the comment on `jobs` says it defaults to the number of cores, but there's no structopt-level default, and here we're checking if the argument was given... are we actually getting parallelism?  Do we need a default in the args struct?

**Review by tjkirch — APPROVED** [review]

File: tough/src/editor/mod.rs (hunk 9 above)
Comment (zmrow) [reply tough/src/editor/mod.rs]: ha good catch - I'll fix it

File: tuftool/src/refresh.rs (hunk 11 above)
Comment (zmrow) [reply tuftool/src/refresh.rs:78]: Good question. Per the [`rayon` FAQ](https://github.com/rayon-rs/rayon/blob/master/FAQ.md), it will spawn the same number of threads as cores. It does this by default when you use any of the parallel iterator methods it provides, which we do below in `build_targets()`.

If a user specifies a specific number of jobs, we use `rayon::ThreadPoolBuilder` to override the default.

Comment (tjkirch) [reply tuftool/src/refresh.rs:78]: Ah, sorry, I was silly and forgot that rayon creates a global thread pool by default.  This is just configuring it, if the user wanted a different number of threads.

Comment (zmrow) [reply tuftool/src/refresh.rs:78]: I'll add a comment for future readers.

File: tough/src/editor/test.rs (hunk 10 above)
Comment (zmrow) [reply tough/src/editor/test.rs:67]: I'll rename this to something else like "file4.txt"

**Comment by zmrow** [comment]
Addressed @tjkirch 's comments

File: tough/src/editor/mod.rs (hunk 9 above)
Comment (tjkirch) [reply tough/src/editor/mod.rs]: That change made me realize that this function doesn't have to return a result - we can always insert into the hash, and so we can remove the burden of error checking from callers.

Comment (zmrow) [reply tough/src/editor/mod.rs]: You're right... tunnel vision over here. I'll fix that before I merge.

**Comment by zmrow** [comment]
Removed the `Result<>` from the `add_target()` method. Good catch @tjkirch !

**Comment by zmrow** [comment]
Fixed the clippy issue.

**Review by tjkirch — APPROVED** [review]
